### DISCUSSION TITLE
What is the point of using the UUID tool in pimcore?

### DISCUSSION THREAD

**Discussion by mhstudioos** [pimcore/pimcore#15673] [Q&A / Support] [answered]
What is the point of using the UUID tool in pimcore?

I'm a symfony developer, and used to work with uuid's for all my entities but with pimcore there is always a integer ID present.

Is the uuid used in the graphql api?

Please give your best advice why/why-not to use the uuid-tool.

**Comment by johanlindgren** [comment]
When working with several environments (staging, production etc) the UUIDs makes it possible to identify the same data with the same UUID even though they will have different entity ids in each environment?

**Reply by NiklasBr** [reply]
This does not appear to be the case, the database ID is used to seed the UUID meaning if the id (`o_id`/`$object->getrId()`) differs from each environment they will get different UUID:s as far as I can see:

https://github.com/pimcore/pimcore/blob/0cee68048c70ac9696c2a8ba195fb26b596a9467/bundles/UuidBundle/src/Model/Tool/UUID.php#L160-L173

Which is then used here:

https://github.com/pimcore/pimcore/blob/0cee68048c70ac9696c2a8ba195fb26b596a9467/bundles/UuidBundle/src/Model/Tool/UUID.php#L129-L145

**Reply by dpfaffenbauer** [reply]
The idea is to have unique IDs when you connect multiple Pimcore Environments. There was, a long time ago, a Bundle to synchronise things between instances. I am not sure if it ever was Open Source or if it was implemented by Elements. But the idea is to have a unique UUID for instance1 and instance2 and then sync them with a custom identifier (like an external ID or something)

**Reply by dpfaffenbauer** [reply]
AFAIK, it used the old Pimcore REST API to sync, so not sure if it ever got ported to DataHub.

**Reply by johanlindgren** [reply]
@NiklasBr correct, but I asume that when syncing objects from stage to prod you would use the uuid already generated for the object in stage as the uuid in prod meaning that the uuid that would have been generated differently in prod would not be used

**Reply by mhstudioos** [reply]
@johanlindgren & @dpfaffenbauer thanks for the reply, does any of you know how to make the uuid accessible in the datahub graphql?

**Reply by johanlindgren** [reply, marked as answer]
Its not part of the systen columns available, that would be a great pullrequest I guess? If you do not have the time for that you could create a calculated value and expose the uuid using a custom calculator class

**Reply by mhstudioos** [reply]
@johanlindgren thanks for sharing your thoughts great help, for now a custom calculator class did the trick (didn't used it before). Creating a pull request won't be able i think because the uuid tool is a separate bundle and not accesible in the datahub bundle. correct me if i'm wrong.
